include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,353 @@
+	int (*disable)(struct k_obj_core *obj_core);
+	int (*enable)(struct k_obj_core *obj_core);
+};
+
+struct k_obj_type {
+	sys_snode_t    node;   /* Node within list of object types */
+	sys_slist_t    list;   /* List of objects of this object type */
+	uint32_t       id;     /* Unique type ID */
+	size_t         obj_core_offset;  /* Offset to obj_core field */
+#ifdef CONFIG_OBJ_CORE_STATS
+	struct k_obj_core_stats_desc *stats_desc;
+#endif
+};
+
+struct k_obj_core {
+	sys_snode_t        node;   /* Object node within object type's list */
+	struct k_obj_type *type;   /* Object type to which object belongs */
+#ifdef CONFIG_OBJ_CORE_STATS
+	void  *stats;              /* Pointer to kernel object's stats */
+#endif
+};
+
+/**
+ * @defgroup obj_core_apis Object Core APIs
+ * @ingroup kernel_apis
+ * @{
+ */
+
+/**
+ * @brief Initialize a specific object type
+ *
+ * Initializes a specific object type and links it into the object core
+ * framework.
+ *
+ * @param type Pointer to the object type to initialize
+ * @param id A means to identify the object type
+ * @param off Offset of object core within the structure
+ *
+ * @retval Pointer to initialized object type
+ */
+struct k_obj_type *z_obj_type_init(struct k_obj_type *type,
+				   uint32_t id, size_t off);
+
+/**
+ * @brief Find a specific object type by ID
+ *
+ * Given an object type ID, this function searches for the object type that
+ * is associated with the specified type ID @a type_id.
+ *
+ * @param type_id  Type ID associated with object type
+ *
+ * @retval NULL if object type not found
+ * @retval Pointer to object type if found
+ */
+struct k_obj_type *k_obj_type_find(uint32_t type_id);
+
+/**
+ * @brief Walk the object type's list of object cores
+ *
+ * This function takes a global spinlock and walks the object type's list
+ * of object cores and invokes the callback function on each element while
+ * holding that lock. Although this will ensure that the list is not modified,
+ * one can expect a significant penalty in terms of performance and latency.
+ *
+ * The callback function shall either return non-zero to stop further walking,
+ * or it shall return 0 to continue walking.
+ *
+ * @param type  Pointer to the object type
+ * @param func  Callback to invoke on each object core of the object type
+ * @param data  Custom data passed to the callback
+ *
+ * @retval non-zero if walk is terminated by the callback; otherwise 0
+ */
+int k_obj_type_walk_locked(struct k_obj_type *type,
+			   int (*func)(struct k_obj_core *, void *),
+				  void *data);
+
+/**
+ * @brief Walk the object type's list of object cores
+ *
+ * This function is similar to k_obj_type_walk_locked() except that it walks
+ * the list without obtaining the global spinlock. Consequently, this routine
+ * does not prevent the addition or removal of object cores while walking the
+ * list. Although this is expected to offer better responsiveness, it may
+ * result in unexpected behavior.
```

Comment:
FWIW I'd be stronger with the prescription here, e.g. "No synchronization is provided here.  Mutation of the list of objects while this function is in-progress must be prevented at the application layer, otherwise undefined/unreliable behavior, corruption, and crashes may result."

Reply:
I like your phrasing. It shall be so. :)